모달 공통 컴포넌트 구조

k-jipgeria opened the pull request:
## #️⃣ 연관된 이슈

- #20 

## 🔨작업 내용

> 이번 PR에서 작업한 내용을 간략히 설명해주세요(이미지 첨부 가능)

- [x] 공통 모달 컴포넌트 구조 구현
- [x] 모달 밖 클릭 시 모달 닫힘
- [x] 모달 사용하는 동안 배경색 어두운 반투명 색으로 설정
- [x] title, mainContent, footer모두 prop로 받음으로써 유연한 개발 가능. 

## 아래는 modal을 열기 위한 과정부터 닫히는 과정의 flow입니다.

### 1. page에서 오픈 모달 버튼 클릭 => `inOpen === true` => 모달 오픈

```ts
export default function TestPage() {
  const [isOpen, setIsOpen] = useState(false);
  const handleClose = () => {
    setIsOpen(!isOpen);
  };
  return (
    <>
      **<button onClick={() => setIsOpen(true)} className="bg-blue-500">
        Open Modal
      </button>**
      {isOpen && <TestModal onClose={handleClose} />}
    </>
  );
}
```

### 2. 임의의 모달에 handleClose(isOpen state 컨트롤 함수) 전달

```ts
const TestModal = ({ onClose }: { onClose: () => void }) => {
  return (
    <GenericModal
      title="Test Modal"
      mainContent={<div>Test Modal</div>}
      footerContent={<div>Test Modal</div>}
      onClose={onClose}
    />
  );
};

export default TestModal;
```

### 3. 공통 모달 컴포넌트(GenericModal.tsx)에 handleClose가 onClose라는 prop으로 전달. => 모달 외 공간 클릭 시 => onClose실행 => 모달 닫힘.

```ts
const GenericModal = ({
  title,
  onClose,
  mainContent,
  footerContent,
}: GenericModalProps) => {
  const handleClose = (e: React.MouseEvent) => {
    if (e.target === e.currentTarget) {
      onClose?.();
    }
  };

  return (
    <div
      className="fixed inset-0 flex items-center justify-center bg-black-000000/30"
      onClick={handleClose}
    >
      <div className="rounded-lg bg-white p-4">
        {title && (
          <header>
            <div>{title}</div>
          </header>
        )}
        <main>{mainContent}</main>
        <footer>{footerContent}</footer>
      </div>
    </div>
  );
};

export default GenericModal;
```

## 버튼 컴포넌트 수정 사항
직접 스타일을 커스텀할 수 있도록 prop으로 커스텀 스타일을 받음.
```ts
const Button = ({
  backgroundColor,
  children,
  customStyle,
}: ButtonProps) => {
  return (
    <button
      className={`${BUTTON_VARIANT_STYLES[backgroundColor]} ${BUTTON_COMMON_STYLES} ${customStyle}`}
    >
      {children}
    </button>
  );
};
```

## 💬 리뷰 요구사항(선택)

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
- 모달에 공통으로 사용하는 컴포넌트라는 점을 고려하셔서 피드백 주시면 좋겠습니다

JIPGERIA committed:
chore: 폴더명을 Modal => Modals로 변경

JIPGERIA committed:
feat: 공통 모달 컴포넌트 GenericModal.tsx 구현 #20

neruu00 commented on app/_components/Modals/GenericModal.tsx:
```diff
@@ -0,0 +1,31 @@
+      </div>
+    </div>
+  )
+}
```
### 컴포넌트 이름이 명확하지 않습니다.
만약 `submit` 이벤트만 다루는 모달인 경우 GenericModal보다는 SubmitModal이 개인적으로 더 명확하다고 생각합니다.

확인 혹은 취소를 다루는 경우는 Confirm을 사용하서도 좋을 거 같아요

### props 명칭
이것 또한 지극히 개인적인 의견이지만 좌측이 흰색에 취소, 오른쪽이 보라색에 제출 이렇게 정해져 있다면 props를 다음과 같이 작명해도 좋을 거 같습니다
- leftButtonText -> closeButtonText
- rightButtonText -> submitButtonText

### 재밌는 아이디어
GenericModal이라면 다음 방법도 고려해 보면 좋을 거 같습니다

모달의 모든 요소를 children으로 받는 방법입니다.

이렇게 하면 버튼도 개발자가 원하는 위치에 원하는 기능을 가진 버튼을 넣을 수 있겠죠?

```ts
//app/_components/modals/GenericModal.tsx
export default function GenericModal({ children }: { children: React.ReactNode }) {
  const main: React.ReactNode[] = [];
  const footer: React.ReactNode[] = [];

  //모달 버튼 컴포넌트면 footer에 아니면 main에
  React.Children.forEach(children, (child) => {
    if (
      React.isValidElement(child) &&
      typeof child.type !== 'string' &&
      child.type.name === 'ModalButton'
    ) {
      footer.push(child);
    } else {
      main.push(child);
    }
  });

  return (
    <div className="fixed inset-0 flex items-center justify-center">
      <div className="rounded-lg bg-white p-4">
        <main>{main}</main>
        <footer className="정렬 스타일">{footer}</footer>
      </div>
    </div>
  );
}
```

```ts
// app/ ... /page.tsx
function Page() {
  return (
    <div>
      <GenericModal>
        <form>...</form>
        <ModalButton theme="purple" handler={handleSubmit}>
          제출
        </ModalButton>
        <ModalButton theme="white" hanlder={handleClickClose}>
          취소
        </ModalButton>
      </GenericModal>
    </div>
  );
}
```

Donghunn-Lee replied:
버튼 명칭에 대해선 저는 `leftButtonText`, `rightButtonText` 의 네이밍이 적절하다고 생각합니다. 보통 헤드리스하게 범용적인 컴포넌트를 구현할 때에는 어떤 노드, 어떤 텍스트가 들어와도 적절히 사용될 수 있도록 하기 때문에 구성된 레이아웃을 기준으로 위치로 네이밍을 하는 방식은 적절하다고 생각합니다. 이 부분에 대해선 다같이 다시 얘기를 나눠보아도 좋겠네요!

Donghunn-Lee replied:
추가로 `children`에 관해 적어주신 부분에선 보통 칠드런은 한 두 가지의 텍스트 보다는 주로 컴포넌트가 들어가는 경우 확장성을 높이기 위해서 사용되는 것으로 알고있습니다. 저희 상단 네비게이션 바에 대쉬보드 멤버를 표시하는 부분을 추가하는 경우에 구성원 컴포넌트를 헤더에 칠드런으로 추가하는 경우를 예로 들 수 있겠네요.

neruu00 replied:
혹시 children의 어떤 의견에 대한 답일까요?

k-jipgeria replied:
@Woolegend 
- 컴포넌트명 SubmitModal.tsx로 변경했습니다.
- ButtonText prop이름 각각 closeButtonText, submitButtonText로 변경했습니다.

k-jipgeria replied:
@Donghunn-Lee 
- ModalButton => LeftButtonText,  RightButtonText로 변경했습니다!

yunrap replied:
여기다 리뷰 연결해서 적을게요! ㅎㅎ
밖에서 chilren을 통해서 버튼컴포넌트를 넣어주는 방향이 확장성면에서 좋을거같다는 생각이듭니다!

버튼이 하나, 아예없거나, 두개이거나 이런 다양한 상황이많을거같네요!

댓글이어서달면 이렇게 남겨지네요 ㅋㅋㅋㅋㅋ

neruu00 commented on app/_components/Modals/GenericModal.tsx:
```diff
@@ -0,0 +1,31 @@
+
+const GenericModal = ({ children, leftButtonText, rightButtonText, onClose, onSubmit }: GenericModalProps) => {
+  return (
+    <div className="fixed inset-0 flex items-center justify-center">
```
이 요소는 모달의 외부 배경인 거 같은데 배경색이나 투명도 혹은 blur 효과를 props로 선택할 수 있으면 좋을 거 같아요

neruu00 commented on app/_components/Modals/ModalButton.tsx:
```diff
@@ -0,0 +1,21 @@
+      {children}
+    </button>
+  )
+}
```
`GenericModal` 컴포넌트에서는 string형태의 text만을 children으로 전달하던데 그 외의 다른 요소도 props로 전달을 목적으로 만드신 걸까요? text만을 children 요소로 사용하려는 경우는 그냥 다른 props로 text만을 명확하게 전달 받는게 더 좋을 거 같아요

아니면 요즘은 로딩 시 스피너를 버튼에 띄워주는 UI도 많던데 이 기능을 계획하신다면 그대로 children으로 받아와도 좋을 거 같아요.

이 리뷰 사항은 그냥 참고만 해주세요!

neruu00 reviewed (changes requested):
린트 설정이 제대로 적용되지 않은 거 같아요. dev 브랜치를 fetch merge 한 후 포맷팅 부탁드립니다!

Donghunn-Lee reviewed (commented):
`props`들의 `?` 옵션 등에 대해서 생각해보게 되는데, 실제로 좀 더 구현해서 적용해보기 전에는 잘 모르겠네요. 지금은 좋은 것 같습니다! 확인했습니다.

yunrap commented on app/_components/Modals/ModalButton.tsx:
```
export default function Button({
  children,
  addClassName,
  handleClick,
  disabled = false,
  type = "button",
  size = "large",
  color = "primary",
  ...rest
}: ButtonProp) {
  const sizeClasses = {
    large: "w-auto desktop:h-[5rem] tablet:h-[4.8rem] tablet:rounded-[1.6rem] ",
    medium:
      "w-auto tablet:h-[5rem] mobile:h-[4.8rem] tablet:rounded-[1.6rem] mobile:rounded-[1.2rem]",
    small: "w-auto rounded-[1.2rem] tablet:h-[4.2rem] mobile:h-[4rem]",
  };

  const colorClasses = {
    primary: "bg-primary hover:bg-[#7b52f1] text-white",
    white:
      "bg-white hover:bg-[#e1d7f4] text-gray-800 border-solid border-[0.1rem] border-gray-300",
  };

  const buttonClass = classNames(
    "rounded transition-all duration-300 ease-in-out cursor-pointer",
    sizeClasses[size],
    colorClasses[color],
    Array.isArray(addClassName) ? addClassName.join(" ") : addClassName,
    {
      "opacity-50 cursor-not-allowed": disabled,
    },
  );

  return (
    <>
      <button
        type={type}
        className={buttonClass}
        onClick={handleClick}
        disabled={disabled}
        {...rest}
      >
        {children}
      </button>
    </>
  );
}
```

버튼 css 를 이렇게 override해서 custom하고 싶은 부분을  addClassName 이렇게 맨마지막요소에 추가해줄수도있는거같습니다!!
랜딩페이지 버튼부분보다가 만들어볼까하다가 지금 css 작업중이라 에잇!하고 우선 넘어갔네요 ㅎㅎㅎ
이렇게하면 추가 클래스가 우선시되서 아마 width하고 heigth는 커스텀하게줄수있을거같은데 ...
한번사용해보고 다시 개인 dm 드려볼게요

k-jipgeria replied:
ModalButton을 다른 분들도 사용할 수 있도록 커스텀 스타일을 prop으로 받도록 수정했습니다

yunrap reviewed (approved):
승인드립니다!

JIPGERIA committed:
refactor: 모달 컨텐츠는 props로 받고, 모달의 틀 역할에만 초점을 맞춤.

- 모든 모달에서 모달 외 공간 클릭 시 모달 Close
- 모든 모달에서 백그라운드 어둡게 설정
- title, mainContent, footerContenet를 모두 props로 받으므로써 유연성 증가

bhoh032019 reviewed (approved):
말씀해주셨던 수정사항이 반영되어 보이네요
추가적으로 수정이 필요한 사항이 보이면 다시 리뷰를 남겨보도록 하겠습니다
고생하셨습니다!!

JIPGERIA committed:
refactor: 공통 버튼을 커스텀할 수 있도록 customStyle을 prop으로 받도록 수정

JIPGERIA committed:
chore: ModalButton 삭제

JIPGERIA committed:
chore: ModalButton 삭제

JIPGERIA committed:
chore: .cursorrules 추적 무시

JIPGERIA committed:
refactore: ModalButton에서 Button으로 컴포넌트 명 변경
